### PR TITLE
[@types/jest] Infer matcher types based on expect

### PR DESCRIPTION
**Pull request by stilettk**
### Description

- Infer matcher types from `expect()`. For example, for `expect(2 + 2).toBe(foo)` the default type for `foo` will be `number` instead of `any`
  - `JestMatchersShape` had to be modified for types to be inferred in `expect().resolves` and `expect().rejects`
  - I have tried `E extends T` instead of `E = T` for more strict type checking (for `expect(2+2).toBe('abc')` to fail), but I have failed to make it work with asymmetric matchers (ex. `expect(myObj).toEqual(expect.objectContaining(...)`)
- Strict type checking for array elements. For example, `expect([1,2,3]).toContain('abc')` will now fail to compile.
  - This is a **breaking change**, but I couldn't find a valid use case that will be broken.
- Strict type checking for `toHaveBeenCalledWith`, `toHaveReturnedWith` and similar
  - Again, this is a **breaking change**, but I couldn't find a valid use case that will be broken. I even had to modify `jest-in-case` tests, but I think they were incorrect, because you have to pass all arguments to `toHaveBeenCalledWith`. And this strict type check helped to find it!
- ~`expect.objectContaining` now accepts `Partial<E>`. So, you can now use `objectContaining<MyType>` instead of `objectContaining<Partial<MyType>>`~ reverted

**Looking for your feedback, especially about the breaking changes**.
- Does it worth it to emit error (see the tests bug found in `jest-in-case`) or it is better to revert to just the default type?
- If breaking changes are ok, where should I increase the version number?

### Checklists

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://jestjs.io/docs/en/expect
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing). _Tested on my work project with 800+ tests_
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by stilettk** [comment]
Wow, I have just fixed a merge ocnflict, and perf warning is now gone 🤔

**Comment by orta** [comment]
Sorry, the Jest types moved out from under you - they _used_ to live in the subfolder, but **now** they live in the root: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47196

Which means you probably want to re-apply your changes to the root index. I think that's why the perf hit went away.

**Comment by stilettk** [comment]
@orta Yep, that's exactly what I did after the merge conflict: I moved my changes from the root (where they used to be) to `ts3.1` subfolder (where they are now). Root `index.d.ts` now has `<reference path="ts3.1/index.d.ts" />`.

**Comment by orta** [comment]
Ah, perfect, sorry, my mis-understanding - cool, everything looks structurally correct

Re: the changes, _to me_, as a Jest user, the breaking changes seem worth it. 

What we do for things like this with Node is wait till the next major version in the JS to ship before making these changes, though it looks like [Jest 27 is quite a while away](https://github.com/facebook/jest/milestone/12). I'll leave the choice on whether/how to bump to the DT module owners though

**Comment by stilettk** [comment]
Honestly, I have no idea why CI has failed on this.
```
Error in react-notifications-component
Error: /home/runner/work/DefinitelyTyped/DefinitelyTyped/types/react-notifications-component/index.d.ts:8:1


ERROR: 8:1  npm-naming  The declaration doesn't match the JavaScript module 'react-notifications-component'. Reason:
The declaration specifies 'export default' but the JavaScript source does not mention 'default' anywhere.
```

**Comment by sandersn** [comment]
Can I get a review from
@geovanisouza92 @NoHomey @jwbay @asvetliakov @alexjoverm @epicallan @ikatyang @wsmd @JamieMason @douglasduteil @ahnpnl @JoshuaKGoldberg @UselessPickles @r3nya @Hotell @sebald @andys8 @antoinebrault @gstamac @ExE-Boss @quassnoi @Belco90 @tonyhallett @ycmjason @devanshj @pawfa @regevbr

**Comment by devanshj** [comment]
I had some thoughts about this type of change [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47553#issuecomment-692248028). I personally have no issues with it, I probably even like it... But I think this is somewhat opinionated and someone from jest should approve of it.

**Comment by stilettk** [comment]
In this case, you can explicitly specify `toEqual<any>(...)`. In other words, it's just a default type.

**Comment by devanshj** [comment]
Hmm didn't think about that, I was thinking maybe one would have to use `// @ts-ignore` which would be very noisy. Passing `any` as a type parameter is not that bad.

File: types/jest/index.d.ts
Comment (devanshj) [inline types/jest/index.d.ts:1030]: We can have something like...
```typescript
declare const expect: <A>(a: A) => ({ toStrictEqual: <E extends ExactWith<A, E>>(e: E) => void })

type ExactWith<A, E> =
  IsPlainObject<A> extends false ? A :
  & { [K in keyof A]: ExactWith<A[K], K extends keyof E ? E[K] : never> }
  & { [K in Exclude<keyof E, keyof A>]: never }

type IsPlainObject<T> =
  T extends object
    ? T extends (...a: any[]) => any
      ? false
      : true
    : false
```
[Playground](https://www.typescriptlang.org/play?ts=true&t=true#code/CYUwxgNghgTiAEYD2A7AzgF3iAHgB3AwC54AeAQQD4AKKE8gSngF5L5qBveDJAZQxgBLMBgCiARwCuUCCVKjsODCBTA08UTigiA6oIwALCgBoNlGiBKimreADckg4PAC+DAFDuMATwIatuvpG5KaibMzu8PAAkmgACtCCKADyAEYAVoQUbLjKquoAZjJoCAD88OTwRJHwAGTwXADaANLwSfAA1iDeSAUVALpWARh6hhQt-aatuSpqnd29GvDlohNV8CggdiAwbC419U2t7ZqQkqCkXT19oqZXi1SDG1s7rp4+frEJUElpmSKkAAq4RqgMUeTmSAyhBqUXKYJm+XYADpUXR4FAUN5Gv0bGxMd5YVFlvAihASkSoiQBJIQESSGSKe5cAQRJwMSQAERQTmuBjInj8IQiCTSCDs9Hc3luZn4QgSrk8vkCvgCYRiKQyBXwKWmVJc1LSjws+VcSVKtwqoXq0Vas0kABMeoNRs8JrZ9p1POdDUQXLA0uVgrVIs14s9uvg+t9YH9gZl7ow2sj0a4sZ1AdcQdVwo1YuT3qjJDTcdMwC5wHjxrlHo5Xs5PpLGarVpDebtdZTxb98AAzK5TJZ4ABGPnuIA)

**Comment by elibarzilay** [comment]
@stilettk / @devanshj -- any conclusion here?

**Comment by devanshj** [comment]
No block from my side, just I think it would be better if we involve someone from jest as it's a significant change.

File: types/jest/index.d.ts
Comment (G-Rath) [inline types/jest/index.d.ts:579]: Why are these being made `Partial`? If I want the properties of the object I'm typing with the generic to be optional, I can use `Partial` myself.

**Review by G-Rath — COMMENTED** [review]
This change looks ok, aside from my comment.

@SimenB would probably be the one to ping from the jest side if you're wanting his input, but it'll probably be along the lines of "if it works, go for it!" 😄

**Comment by elibarzilay** [comment]
OK, so this is now waiting for @stilettk for the comment and for @SimenB to say something...

(BTW, sorry for the long time this went on, it's due to changes in the pr workflow that we did recently.  See DefinitelyTyped/dt-mergebot#204.)

**Comment by SimenB** [comment]
As expected, I think this is a great idea 👍 Whether this should wait for Jest 27 or not I don't really have an opinion about.

---

We (jest core) really need to start porting over these type improvements at some point.

**Comment by elibarzilay** [comment]
Down to just @stilettk now--?

**Comment by G-Rath** [comment]
@elibarzilay I'd like [my comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020#discussion_r510720375) addressed/discussed before this is merged.

File: types/jest/index.d.ts
Comment (stilettk) [reply types/jest/index.d.ts:579]: I made this change because I can see no reason for `objectContaining` to accept entire object. From [jest docs](https://jestjs.io/docs/en/expect#expectobjectcontainingobject): 

> That is, the expected object is a **subset** of the received object

If you need to check for entire object, why would you use `.toBeCalledWith(expect.objectContaining(value))` instead of just `.toBeCalledWith(value)`?

Comment (G-Rath) [reply types/jest/index.d.ts:579]: Being a subset doesn't mean it doesn't require properties - Using `Partial` would make sense if `E` was the object being passed in to `expect`, but that's not required.

Consider this:

```
interface CommonAttributes {
  id: string;
  createdAt: string;
  modifiedAt: string;
}

it('contains the common attributes', () => {
  expect(model.attributes).toStrictEqual(
    expect.objectContaining<CommonAttributes>({
      id: expect.any(Number),
      createdAt: expect.any(String),
      modifiedAt: expect.any(String)
    })
  );
});
```

That is testing that `model` is a subset of `CommonAttributes`, which in order to be true requires specific properties.

Comment (stilettk) [reply types/jest/index.d.ts:579]: Thanks for the example, it really makes sense. However I think that this is rather unusual case - in my experience most usages are like `expect.objectContaining<Partial<MyInterface>>({...})`, which seems kinda cumbersome. In your case I would use `expect(model.attributes).toMatchObject<CommonAttributes>(...)` - this one is unchanged.

If there are any other examples, I agree that this particular change should be reverted. Additionally, one can argue that even this example can be considered as unnecessary breaking change.

Please tell me what you think.

Comment (G-Rath) [reply types/jest/index.d.ts:579]: My example is a simplification of a pretty reasonable one - it's about testing duck-typing, which is actually really powerful, since it lets you leverage TypeScript to improve testing that things are of the right subtype (because i.e if I suddenly decide to require all my models need to have a `slug` as well, TS will tell me everywhere I need to update my tests).

[`toMatchObject` is not a direct replacement for `expect.objectContaining`](https://stackoverflow.com/questions/45692456/whats-the-difference-between-tomatchobject-and-objectcontaining):

```
  // objectContaining, with nested object, containing partial props/values
  // FAILS
  expect({ position: { x: 0, y: 0 } }).toEqual(expect.objectContaining({
    position: {
      x: expect.any(Number)
    }
  }));

  // toMatchObject, with nested object, containing partial props/values
  // PASSES
  expect({ position: { x: 0, y: 0 } }).toMatchObject({
    position: {
      x: expect.any(Number)
    }
  });
```

[It's even been mentioned the possibility of removing it](https://github.com/jest-community/eslint-plugin-jest/issues/230#issuecomment-469253911) in favor of using `expect.objectContaining` (afaik the only reason its still there, aside from removing a matcher being a huge breaking change, is that it's a little extra to type `expect.objectContaining` + a matcher).

> Additionally, one can argue that even this example can be considered as unnecessary breaking change.

I'm not sure what you mean by this.

-----

At the very least, I don't think this change should be made in this PR as it's not required for it and should be considered separate.

I understand why this seems like a reasonable change, and agree it'd be nice, but my issue is **this prevents you from ever marking a property as required**, where the gain is just saving ~5 characters by not typing `Partial` yourself.

File: types/jest/index.d.ts
Comment (stilettk) [reply types/jest/index.d.ts:1030]: Sorry for misunderstanding, but what problem does this code solve? As I can see from the playground, it checks for additional properies, so `expect({ a: "a" }).toStrictEqual({ a: "a", b: "b" })` will fail. But can't this be achieved with just `toStrictEqual(expected: T): R`?

Also, your example is too restrictive. Consider this example:
```
interface Animal { name: string }
interface Dog extends Animal { breed: string }
const dog: Dog = { name: 'Rex', breed: 'chihuahua' }
const animals: Animal[] = [ dog ]
expect(animals[0]).toStrictEqual({ name: 'Rex', breed: 'chihuahua' })
```

Comment (stilettk) [reply types/jest/index.d.ts:1030]: Sorry for misunderstanding, but what problem does this code solve? As I can see from the playground, it checks for additional properies, so `expect({ a: "a" }).toStrictEqual({ a: "a", b: "b" })` will fail. But can't this be achieved with just `toStrictEqual(expected: T): R`?

Also, your example is too restrictive. Consider this example:
```
interface Animal { type: string }
interface Dog extends Animal { breed: string }
const dog: Dog = { type: 'dog', breed: 'chihuahua' }
const animals: Animal[] = [ dog ]
expect(animals[0]).toStrictEqual({ type: 'dog', breed: 'chihuahua' })
```

File: types/jest/index.d.ts
Comment (stilettk) [reply types/jest/index.d.ts:579]: Thank you for your explanation! I'll remove this change and add a test for it 👍

**Comment by elibarzilay** [comment]
@G-Rath, yeah, that's what I was referring to (which is perhaps the update from a few minutes ago)...

File: types/jest/index.d.ts
Comment (devanshj) [reply types/jest/index.d.ts:1030]: (afk) your version won't check for deep equalities and mine would as the playground shows.
Although as you pointed out the actual runtime value can be wider than the type hence we shouldn't have my version. Apart from that I think this would be only work well for literals in other cases I feel it would be a footgun to it's not a great idea anyway. Probably I was a little drunk with type gymnastics when I suggested this xD.
So tbc I'm taking back this suggestion.

Comment (devanshj) [reply types/jest/index.d.ts:1030]: (on keyboard now) I know this is resolved by realized this part - "your version won't check for deep equalities and mine would" - is not clear in the above playground. So here's a version which would show the difference.

```typescript
declare const expect: <A>(a: A) => ({
  toStrictEqualMine: <E extends ExactWith<A, E>>(e: E) => void,
  toStrictEqualYours: (e: A) => void,
})

type ExactWith<A, E> =
  IsPlainObject<A> extends false ? A :
  & { [K in keyof A]: ExactWith<A[K], K extends keyof E ? E[K] : never> }
  & { [K in Exclude<keyof E, keyof A>]: never }

type IsPlainObject<T> =
  T extends object
    ? T extends (...a: any[]) => any
      ? false
      : true
    : false

let a = { a: "a", b: { c: "c" } }
let b = { a: "a", b: { c: "c", e: 1 } }

expect(a).toStrictEqualMine(b) // error
expect(a).toStrictEqualYours(b) // no error
```

[Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3iAHgB3AwC54AeAQQD4AKKE8gSngF5L5qBvAKHngyQDKGGAEswGAKIBHAK5QIAWREoQJUhOw4MIFMDTwJOKOIDqIjAAsKAGgOUaqg01bwAbkhHBrPPoOFjJWXkATSQZGDQSakdGFjZ3T28AXwYuLgwATwIDI1NzK3JbCTZmHwBJNAAFaGUAeQAjACtCCjZcbV19ADN5NAQAfnhyeCIfADJ4DngAbQBpeGV4AGsQDKQuoYBdEkNjDDNLCjnN23n2nT1l1fWDeEGJY5H4FVcQGDYk8cmZ+cXDSBkoFIKzWGwkthBNyo22eIFeMHgn3SWQQFWqUDqTRaABUSj5sZoOpckFjxD5eIMCedOuwAHT0ujwKAoDLTTbONjMjLk3h3eA9CB9Hm8EjCGQgHkkAVCrgQEBYKAsb6MgBEUBVtnqJCmYBIKrAKsRiNl8vg9SVU1V6s12sQeoNtkcAEYjUjcARxLQGLT+EJROJpHJFMoQNR6kwAPQR7AwGBIGBcd2EL0+vz+wJB0LhNBhyPRlBIGNxmBAA). The previous version doesn't show the difference probably because they were literals and/or because of contextual inference idk.
Anyway, I just wanted to clarify how my version is different. Still isn't workable.

**Review by RyanCavanaugh — CHANGES_REQUESTED** [review]
Leaving a more clear block on this per @G-Rath's comment

**Comment by stilettk** [comment]
@RyanCavanaugh the changes requested by @G-Rath are already addressed in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020/commits/e9bd7cbe94dd4ce5b378618e03901cbc4bc05088.

**Comment by stilettk** [comment]
I have already made the required changes in https://github.com/DefinitelyTyped/DefinitelyTyped/commit/e9bd7cbe94dd4ce5b378618e03901cbc4bc05088. I'm waiting for review from @RyanCavanaugh.

**Comment by sandersn** [comment]
@G-Rath I think you are the only one left to sign off on this (@RyanCavanaugh just requested changes so the bot would understand the current review state.)

**Review by G-Rath — APPROVED** [review]
LGTM 👍

**Comment by ujwal-setlur** [comment]
This just broke ```expect().toHaveBeenCalledWith()```

```bash
expect(next).toHaveBeenCalledWith(ctx);
tests/middlewares/moleculer.log.middleware.spec.ts:54:5 - error TS2554: Expected 0 arguments, but got 1.
```

**Comment by pkuczynski** [comment]
I can confirm this breaks `toHaveBeenCalledWith`. In my case:

```
const logger = winston.createLogger(...)
const log = jest.spyOn(logger, 'log')
 expect(log).toHaveBeenCalledWith('info', ..., { meta1, meta2 })
```

Produces TS error: `expected 2 arguments instead of 3`

@ujwal-setlur how did you fixed it in your case? Can you provide an example?

@stilettk any advice on how to fix my example above? I tried with `const log = jest.spyOn<Logger, 'log'>(logger, 'log')` but it kept failing.

**Comment by ujwal-setlur** [comment]
@pkuczynski, here is what I did:

Before:

```ts
const log = jest.fn(() => {});

expect(log).toHaveBeenCalledWith('Hello World'); // <-- This now gives TS error
```

After:

```ts
const log = jest.fn((msg: string) => msg);

expect(log).toHaveBeenCalledWith('Hello World'); // <-- This now works since we have specified the type of the argument

```

**Comment by ujwal-setlur** [comment]
@pkuczynski, try something like:

```ts
const log = jest.spyOn(logger, 'log').mockImplementation((arg1: string, arg2: string) => {});
```

You will likely need to specify the correct number of args and their types...

**Comment by Leksat** [comment]
In case of overloaded functions, it takes the last declaration.

Noticed it when tried to spy on `extract`:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8d0c96991365c7b862fa341fd93dbd0b498fe596/types/tar/index.d.ts#L690-L704

My code:
```ts
const extractSpy = jest.spyOn(tar, 'extract');
expect(extractSpy).toHaveBeenCalledWith({
  sync: true,
  file: 'test/test-1.0.1.tar.gz',
  cwd: 'test',
});
```

Error:
```
TS2554: Expected 3 arguments, but got 1.
```

But if I edit `tar/index.d.ts` and move the first function declaration to the last position, the error is gone.

**Comment by Leksat** [comment]
BTW, is there any way in TS to pick a certain declaration of an overloaded function? E.g. to reuse it in custom code.

**Comment by lallenlowe** [comment]
I am surprised by this. This is supposed to be a minor change, and it is wreaking unholy havoc with our tests.

**Comment by ujwal-setlur** [comment]
Yes, this should have been published as a major revision change.

**Comment by rbadr** [comment]
Hello @stilettk @G-Rath 👋

Can you please release a major version for this breaking change ? 
 
It's currently breaking everything. 
Thank you 🙏

**Comment by ujwal-setlur** [comment]
Also, how to address @Leksat  issue of overloaded functions? Reordering is not a sustainable solution.

**Comment by ExE-Boss** [comment]
> If breaking changes are ok, where should I increase the version number?

The `major` and `minor` versions of a `@types` package must match the `major` and `minor` versions of the package that is typed by the `@types` package, so this should probably be reverted until **Jest v27** is released.

**Comment by doberkofler** [comment]
This (already published) PR breaks about 10% of all my tests. Especially the new type inferred version of `toStrictEqual` makes it almost impossible to be used. I do not understand why a function that is by design supposed to catch problems at runtime should need to be typed and fail at compile time? At least for me, this change makes the new type definition extremely unpractical and would force me to cast a very large number of arguments that have a "wrong" type on purpose. I strongly recommend to switch back to a simple `any` type.

**Comment by lallenlowe** [comment]
We ended up casting to `unknown` in a large number of tests to cheese our way past this. like: `expect(results as unknown).toContainEqual(foo)`

**Comment by doberkofler** [comment]
@lallenlowe This is an option but seems a lot of (in my opinion quite useless casts) to type check something that does not need to be checked. I really love a properly typed production code and the stricter the better, but not when I'm writing unit tests that should test my code per definition. Maybe someone could explain the rationale why this "feature" was added in the first place?

**Comment by stilettk** [comment]
Hi everybody!

I agree that this should not have been released in a minor version. Honestly, I don't know what to do now because I don't know the rules of DefinitelyTyped community. @sandersn can you help please? What to do now?

Also, let me explain the reasoning of this PR. Once I made a refactoring in my code, and forgot to change one of the usages, and it introduced a runtime error. My tests didn't catch this error, because the tests were still checking for the *old* object structure. If a type was specified, I would get a compile time error in my test and have fixed it.
Now every single time I specify a type when calling `expect(...).toContainEqual<MyType>(...)`. That's why I created this PR - to make life easier.

I tried to find a case when it will be wrong - every case I could think of was because of the wrong typings. Some examples in this thread can indeed be fixed by correcting the typings: for example, `jest.fn(() => {})`->`jest.fn()` will work, there is even [a test for it](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020/files#diff-ff7b882e4a29e7fe0e348a6bdf8b11774d606eaa221009b166b01389576d921fR968). But the cases like `jest.spyOn` and overloads really seem like a bug. I didn't investigate it yet so I can't say if this is a this a `spyOn` typings problem or a bug in the PR. But I'm sure that cases like `spyOn` should be added to tests.

**Comment by pkuczynski** [comment]
I think the biggest issue is what @Leksat mentions: it breaks heavily on overloadeds. In my case:

```js
const logger = winston.createLogger(...)
const log = jest.spyOn(logger, 'log')
expect(log).toHaveBeenCalledWith('info', ..., { meta1, meta2 })
```

Produces TS error: `expected 2 arguments instead of 3`, because `logger.log` is typed as:

```js
interface LogMethod {
    (level: string, message: string, callback: LogCallback): Logger;
    (level: string, message: string, meta: any, callback: LogCallback): Logger;
    (level: string, message: string, ...meta: any[]): Logger;
    (entry: LogEntry): Logger;
    (level: string, message: any): Logger;
  }

interface Logger extends NodeJSStream.Transform {
  // ...
  log: LogMethod;
}
```

So now TS thinks that `LogMethod` is actually `(level: string, message: any): Logger;`, while in my code I am calling the other version `(level: string, message: string, ...meta: any[]): Logger;`. I don't know any method to suggest TS which overload I am planning to use :(

**Comment by ujwal-setlur** [comment]
I suggest this change gets reverted and another update is released. Then this can get revisited for a more major release.

**Comment by pkuczynski** [comment]
I agree

**Comment by stilettk** [comment]
Found it - this is not a `spyOn` issue, it's the design limitation: https://github.com/microsoft/TypeScript/issues/26591#issuecomment-414894092
I definitely want this to be added as a test.

**Comment by pkuczynski** [comment]
Does not seem it would be solved soon by TS, so how would you propose to solve this situation?

**Comment by stilettk** [comment]
Well, aside from fixing the types version and waiting for a revert - I can suggest you to specify types explicitly, like `const log = jest.spyOn(logger, 'log') as jest.SpyInstance`

**Comment by doberkofler** [comment]
@stilettk Sorry for repeating myself, but I really believe that the change should be complete reverted and the API should allow to compare different types. In my opinion not the specific `jest.spyOn` restriction is causing a lot of pain but rather the much larger number of used jest matchers are the real pain point here.

Given the fact that I use this API to detect runtime errors, I find it quite disturbing, that the subtle type differences in a test like the following report an error and must be type casted. I believe this stricter type checking does cause more harm then it does good:

```typescript
const c: 1|2 = 1;
const e = 3 - 1;
expect(c).toStrictEqual(e); // perfectly valid test throws a type error
expect(c).toStrictEqual(e as unknown as 1|2) // needed type cast to compile
```

Why should I have to refactor a ton of perfectly valid unit tests, just to stay on the latest version of the type definitions?

**Comment by pkuczynski** [comment]
I ended up pin-pointing this dependency `"@types/jest": "26.0.15"`. @stilettk any plans to revert this soon and release new version?

**Comment by stilettk** [comment]
@pkuczynski Hello! I agree that this should be reverted, but I thought that this can be done only by maintainers?

**Comment by pkuczynski** [comment]
I think you can always create a new PR that reverts your changes? This might be the fastest...

**Comment by Shinigami92** [comment]
```ts
const expected: ReadonlyArray<string> = 'aedcfghilnb'.split('');
const code: string[] = 'aedcfghilnb'.split('');
const actual: string[] = stableSort(/*...,code,...*/);

expect(actual).toStrictEqual(expected);
```

image

Was totally valid before 😕

**Comment by doberkofler** [comment]
Sorry to insist, but will someone now revert this breaking change or are we stuck with this very unpractical API and will have to change all of our broken unit tests?

**Comment by pkuczynski** [comment]
Here you guys https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50045. I guess someone has to approve it now and merge...

**Comment by pkuczynski** [comment]
@types/jest@26.0.19 published to npm reverts this change. Check #50045 for more details.

**Pull request merged**
